### PR TITLE
Updated the generator phpdoc to help identifier magic methods

### PR DESCRIPTION
**Pull request by pimjansen**
### What is the reason for this PR?

As mentioned in #303 there are some problems with static code analysis for our deprecation warnings in 1.14. With this PR i did some reflection and generated a new PHPDoc to match the actual code itself

- [ ] A new feature
- [ ] Fixed an issue (resolve #ID)

### Author's checklist

- [x] Follow the [Contribution Guide](https://github.com/FakerPHP/Faker/blob/main/.github/CONTRIBUTING.md)
- [ ] New features and changes are [documented](https://github.com/FakerPHP/fakerphp.github.io)

### Review checklist

- [ ] All checks have passed
- [ ] Changes are approved by maintainer

### CODE REVIEW

**Comment by pimjansen** [comment]
For the ones that get worse, this is due the fact that no return type has been specified on the methods itself. Do we want to drag this into scope or not? If yes then we need to modify those blocks as well

**Comment by pimjansen** [comment]
@localheinz can you assist in the weird phpstan error? Can't resolve it somehow. Also when running it locally im getting phpstan errors.

```
Note: Using configuration file /home/pim/Projects/Faker/phpstan.neon.dist.
 478/478 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

 -- -------------------------------------------------------------------------------------------------------------------- 
     Error                                                                                                               
 -- -------------------------------------------------------------------------------------------------------------------- 
     Reached internal errors count limit of 50, exiting...                                                               
     Internal error: Internal error: preg_replace(): Argument #3 ($subject) must be of type array|string, null given in  
     file /home/pim/Projects/Faker/src/Faker/Provider/hr_HR/Address.php     
```

Could be the fact that its running php 8 locally though.

**Review by bramceulemans — COMMENTED** [review]
```
Ignored error pattern #^Return typehint of method Faker\\Provider\\me_ME\\Address\:\:localCoordinates\(\) has invalid type Faker\\Provider\\latitude\.$# in path /home/runner/work/Faker/Faker/src/Faker/Provider/me_ME/Address.php was not matched in reported errors.
```

Maybe that error should be fixed

**Review by bramceulemans — APPROVED** [review]

File: src/Faker/Provider/Address.php [hunk 1]
```diff
@@ -132,7 +154,7 @@ public static function longitude($min = -180, $max = 180)
     /**
      * @example array('77.147489', '86.211205')
      *
-     * @return array | latitude, longitude
+     * @return float[]
```
Comment (localheinz) [inline src/Faker/Provider/Address.php:157]:
```suggestion
     * @return array{0: float, 1: float}
```

File: src/Faker/Provider/DateTime.php [hunk 2]
```diff
@@ -353,6 +353,8 @@ private static function setTimezone(\DateTime $dt, $timezone)
      * Sets default time zone.
      *
      * @param string $timezone
+     *
+     * @return void;
```
Comment (localheinz) [inline src/Faker/Provider/DateTime.php:357]:
```suggestion
```

**Comment by localheinz** [comment]
@pimjansen 

Can you verify whether you have Xdebug enabled? Disabling it should fix the error.

As @bramceulemans suggested, you should be able to regenerate the baseline afterwards!

**Comment by localheinz** [comment]
@pimjansen 

What do you think about running the script that updates the DocBlock in GitHub Actions?

This way we could ensure that

- the script works 
- the script is run when changes were made

**Comment by pimjansen** [comment]
> @pimjansen 
> 
> 
> 
> What do you think about running the script that updates the DocBlock in GitHub Actions?
> 
> 
> 
> This way we could ensure that
> 
> 
> 
> - the script works 
> 
> - the script is run when changes were made

Could do but maybe its a bit too much magic and it reduces the actual understanding what everything does imo.

Concerning the issue, the error is a bit odd since the method actually is no phpdoc defined that gives the error. But let me run without xdebug

**Comment by pimjansen** [comment]
@localheinz seems that PHPstan version is not running fine on PHP8. Lets modify that in the future. I will downgrade locally to 7.4 instead where it runs fine

**Comment by lefuturiste** [comment]
Do you known when this will be released ?

**Comment by pimjansen** [comment]
Unknown yet. Since its not a bug we are not pushing an extra release for this

**Comment by Pixelshaped** [comment]
It's still pretty inconvenient. Please consider it, unless you're close to your next release. Also you should probably not close issues like #303 until this is merged into main: it gives the false impression the issue is fixed while it isn't. You can use closing keywords in your PR description like "Fix #303" for it to automatically happen upon merge.

**Comment by stof** [comment]
@pimjansen any chance to tag a release including that fix ? It's been more than 2 months since the fix was done now.
Without this fix, we must choose between running static analysis and migrating away from the deprecated properties, but we cannot have a nice state.

**Comment by pimjansen** [comment]
> @pimjansen any chance to tag a release including that fix ? It's been more than 2 months since the fix was done now.
> Without this fix, we must choose between running static analysis and migrating away from the deprecated properties, but we cannot have a nice state.

https://github.com/FakerPHP/Faker/pull/342

**Comment by stof** [comment]
There's one thing that got worse in this update: `unique()` and `optional()` have been updated to document their return type as `UniqueGenerator` and `DefaultGenerator`. But those 2 classes don't have the magic methods documented on them.
Previously, the magic methods were documented as returning `Generator`

**Comment by shalvah** [comment]
Hey, something else is broken here. Was using the `shuffle()` method to shuffle a string. Worked fine in v14, but v15 gives an error with PHPStan, because the return type was changed from `array|string` to `array`, which is incorrect. I'm guessing it was a mistake?

**Pull request merged**